README.md
jQuery Chaos Modal
==================

A Simple yet extendable jQuery modal script built for use with inline HTML, forms, and images in mind.

There are many other modal plugins out there. The goal of this project is to make one that is dead simple to use and implement, while still allowing for easy customization.

Example Usage
-------------

```
<html>
	<head>
		<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"></script>
		<script src="jquery-modal.js"></script>
	</head>
	<body>
		<h1>jQuery Chaos Modal Examples</h1>
		<h2>Image modals</h2>
		<a href="test_image.png" class="chaos-modal-link"><img src="test_image.png" /></a>
		<div class="chaos-modal-link">
			<a href="test_image.png"><img src="test_image.png" /></a>
		</div>
		<h2>Html Content Modals</h2>
		<div class="content-wrap">
			<a href="#" class="chaos-modal-link">Click Here to open the modal</a>
			<div class="chaos-modal-box">
				All HTML in here is used as the modal's content.
			</div>
		</div>
		<h2>Html Content ID Modals</h2>
		<div class="content-wrap">
			<a href="#" class="chaos-modal-link" data-chaos-modal-box-id="modal-content-id">Click Here to open the modal</a>
		</div>
		<div class="modal-content-id">
			All HTML in here is used as the modal's content.
		</div>
	</body>
</html>
```

Events Usage
----------------
```html
<html>
	...
	<!-- Put data-preprocess='true' on link of modal you wish to pre-process -->
	<div class='chaos-modal-link' data-preprocess='true'>Link</div>
	...
</html>
```

```javascript
var modal_links = $('.chaos-modal-link');

//Runs on modal link selector before modal is displayed
//Modal variable contains the modal associated with the modal link.
modal_links.on('chaos-modal-preprocess', function( event, modal){

	$(window).on('chaos-modal-preprocess-interrupt', function( event ){
		//Event fires when modal/loading screen is closed.
	});
	
	//If you do not want to pre-process every time, set data-preprocess to false
	//When done pre-processing
	$(this).attr('data-preprocess','false');
	
	//Preprocessing function MUST call openModal when preprocessing complete
	modal.openModal();
});
```

Options Usage
----------------
Attach data attributes to the modal's link as so;

```html
<a class="chaos-modal-link" data-chaosmodal-max-width="710"> Link </a>

```
###Supported options:
####data-chaosmodal-max-width : 
Sets the maximum width of the modal
####data-chaosmodal-close-link : 
Sets whether or not to display default close link


Features
--------

* Stand alone JS file. No additional CSS file required.
* Window resize handling for better performance on responsive websites.
* Vertical scroll position retention to maintain the user's position on the webpage.
* On-the-fly modal HTML generation. No preprocessing loops before the page is shown to the user.
* Content type handling:
	* Image tags
	* Links to images
	* Inline HTML
	* Inline HTML defined by element id
* Iframe lazy loading for iframes in hidden modal content to improve page load.
* Internet Explorer support back to version 8.
* Microsoft Edge support.
* Event support for pre-processing the modal content.
* Options for changing default behavior.

Wishlist
--------

* Options for changing the CSS styles.

CHANGELOG.md
Change Log
==========

Recent changes
--------------
* Fixed iframe security bug in MSEdge (1.5.0)
* Added option to set maximum width on modals (1.4.2)
* Added option to determine whether or not to print out default close link (1.4.2)
* Fixed iframe security bug in IE (1.4.1)
* Added JS events for pre-processing of modal content (1.4)
* Improved image and iframe content loading (1.4)
* Added support for JS events in modal content (1.3.10)
* Fixed IE8 bug that caused the resize event to continuously fire during the show animation (1.3.10)
* Fixed IE bug that prevented modal content from showing under certain conditions (1.3.10)
* Fixed bug in iframe lazy load where data-src attribute was not being removed (1.3.9)
* Converted all global variables to jQuery properties (1.3.8)
* Improved Iframe content handling (1.3.8)
* Iframes inside of modal box content are now lazy loaded if the modal box content is hidden on the page (1.3.8)
* Improved Image width handling (1.3.7)
* Fixed bug related to undefined data attributes (1.3.7)
* Added defining of modal content by id in data attribute (1.3.6)
* Corrected clone content handling to better respect element ids (1.3.6)
* Corrected typo in JSON manifest (1.3.5)
* Added manifest file for release on plugins.jquery.com (1.3.4)
* Added handling of link class applied to a link's direct container (1.3.3)
* Provided a minifed version of the script (1.3.2)
* Moved Change Log to its own file to decrease the file size (1.3.2)
* Fixed closeOnESC not passing its event to closeCurrentModal (1.3.1)
* Fixed container styles on custom html (1.3.1)
* Fixed improper class name on custom html containers (1.3.1)
* Added alwaysAtTop parameter to openModal and resizeModal function (1.3.0)
* Implemented vertical positioning from current scroll position (1.3.0)
* Fixed browser following the href of print and close links (1.3.0)
* Added image prerendering in modal content (1.2.0)
* Updated and fixed resizeModal function to work based off of the html element's width (1.2.0)
* Added image in link href support (1.2.0)
* Reworked automatic modal binding to clone content on-the-fly (1.2.0)
* Added maxWidth parameter to openModal function (1.1.0)
* Added automatic modal binding to .chaos-modal-link elements (1.1.0)
* Added option to close the modal when the mask is clicked or the ESC key is pressed (1.1.0)
* Reworked references to optimize javascript caching (1.1.0)
* Clones content container and appends to body for absolute positioning relative to body
* Added max width variable so you can keep modal content inside your wrapper div's width

Things left to do
-----------------
* Function parameters for changing the default CSS styles
	* Mask color
	* Mask opacity
	* Default padding
	* Content background
* Jquery animation queue integration
* Change namespacing of functions to work similar to $('#example').modal("open")
* Add position parameter to openModal function

jquery-modal.js
/*
 * jQuery Chaos Modal
 * By Matthew Sigley
 * Based on concept work by Kevin Liew - http://www.queness.com/post/77/simple-jquery-modal-window-tutorial
 * Version 1.3.1
 * Recent changes:
 * - Fixed closeOnESC not passing its event to closeCurrentModal (1.3.1)
 * - Fixed container styles on custom html (1.3.1)
 * - Fixed improper class name on custom html containers (1.3.1)
 * - Added alwaysAtTop parameter to openModal and resizeModal function (1.3.0)
 * - Implemented vertical positioning from current scroll position (1.3.0)
 * - Fixed browser following the href of print and close links (1.3.0)
 * - Added image prerendering in modal content (1.2.0)
 * - Updated and fixed resizeModal function to work based off of the html element's width (1.2.0)
 * - Added image in link href support (1.2.0)
 * - Reworked automatic modal binding to clone content on-the-fly (1.2.0)
 * - Added maxWidth parameter to openModal function (1.1.0)
 * - Added automatic modal binding to .chaos-modal-link elements (1.1.0)
 * - Added option to close the modal when the mask is clicked or the ESC key is pressed (1.1.0)
 * - Reworked references to optimize javascript caching (1.1.0)
 * - Clones content container and appends to body for absolute positioning relative to body
 * - Added max width variable so you can keep modal content inside your wrapper div's width
 * Things left to do:
 * -Add padding option to openModal
 * -Jquery animation queue integration
 * -Function parameters for changing the default CSS styles
 * -Change namespacing of functions to work similar to $('#example').modal("open")
 * -Add position parameter to openModal function
 */

(function( $ ) {
	//Property to store the modal current being displayed
	var currentModal,
		maxWidth;
	
	//Global variable for generating modal box clones
	$.chaosModalIndex = 0;
	
	$.fn.openModal = function( options ) {
		//Default options
		var defaults = { maskClose: true, maxWidth: 960, closeLink: true, printLink: false, alwaysAtTop: false };
		options = $.extend({}, defaults, options);
		
		//Clone modal content
		var clone = this.clone(),
			bodyElement = $('body'),
			htmlElement =  $('html'),
			documentElement = $(document),
			windowElement = $(window),
			maskClose = options['maskClose'],
			closeLink = options['closeLink'],
			printLink = options['printLink'],
			alwaysAtTop = options['alwaysAtTop'];
			
		//Update maxWidth
		maxWidth = options['maxWidth'];
		
		//Update currentModal
		currentModal = clone;
		clone.attr({id: 'chaos-current-modal'});
		
		//Write the mask div
		var modalMask = $('<div id="chaos-modal-mask" class="chaos-modal-mask"></div>');
		modalMask.css({'position': 'absolute', 'z-index': '9000', 'background-color': '#000', 'display': 'none', 'top': '0', 'left': '0'});
		modalMask.appendTo(bodyElement);
		
		//Write print link if none exist
		if(clone.find('.print-link').length == 0 && printLink) {
			clone.prepend('<a class="print-link">Print</a>');
			clone.children('.print-link').css({'float': 'right', 'margin': '5px'});
			}
		
		//Write close link if none exist
		if(clone.find('.close-link').length == 0 && closeLink) {
			clone.prepend('<a class="close-link">Close</a>');
			clone.children('.close-link').css({'float': 'right', 'margin': '5px'});
			}
               
        //Set the popup window css
        clone.css({'display': 'block', 'position': 'absolute', 'background': '#fff', 'z-index': '9001', 'left': '-10000px', 'margin': '0', 'padding': '0'});
        clone.appendTo(bodyElement);
        
        var showModal = function() {
        	//Lock popup window width
	       	if(clone.width() > maxWidth){
	       		clone.width(maxWidth);
	       	} else {
	       		clone.width(clone.width());
	       	}
	       	clone.hide();
	       	
	     	//Calculate the modal mask size and popup window position
	     	clone.resizeModal( alwaysAtTop );
	     	
	     	//Mask transition effect    
	        modalMask.fadeIn(1000);   
	        modalMask.fadeTo("slow",0.8);
	     	
	        //transition effect
	        clone.fadeIn(2000);
	        
	        //Bind the window resize event
	        windowElement.bind('resize', {'alwaysAtTop': alwaysAtTop}, resizeCurrentModal);
	        
	        //Bind the print link events if any close links exist
	        if(clone.find('.print-link').length > 0) {
	        	clone.find('.print-link').bind('click', printCurrentModal);
	        	}
	        
	        //Bind the close link events if any close links exist
	        if(clone.find('.close-link').length > 0) {
	        	clone.find('.close-link').bind('click', closeCurrentModal);
	        	}
	        
	        //Bind close event to mask click and when the ESC key is pressed
			if(maskClose) {
				modalMask.on('click', closeCurrentModal);
				documentElement.on('keyup', closeOnESC);
			}
        };
        
        //Show modal after all images have loaded
        var lastImage = clone.find('img').last();
        if(lastImage.length) { lastImage.on('load', showModal); }
        else { showModal(); }
        
        clone.show();
        
        return this;
	};
	
	$.fn.closeModal = function() {
		//Clear currentModal
		currentModal = null;
		
		//Unbind the window resize event
		$(window).unbind('resize', resizeCurrentModal);
		
		//Unbind the close link events if any close links exist
        if(this.find('.close-link').length > 0) {
        	this.find('.close-link').unbind('click');
        	}
		
		//Remove the popup window
		this.hide().remove();
		
		//Remove the mask div
		$('#chaos-modal-mask').hide().remove();
		
		return this;
	};
	
	$.fn.resizeModal = function( alwaysAtTop ) {
		var documentElement = $(document),
			windowElement = $(window),
			htmlElement = $('html'),
			maskHeight = documentElement.height(), //Get the screen height and width
			maskWidth = htmlElement.width(),
			winH = windowElement.height(), //Get the window height and width
        	winW = windowElement.width();
        	
        //Check for invalid mask dimensions
        if(maskHeight < this.height()) { maskHeight = this.height(); }
        if(maskWidth < winW) { maskWidth = winW; }
        
        //Set height and width to mask to fill up the whole screen
        $('#chaos-modal-mask').css({'width':maskWidth,'height':maskHeight});
                
        //Calculate popup window position
        var modalTop = winH/2-this.height()/2,
        	modalLeft = winW/2-this.width()/2;
        
        //Check for invalid window positions
        if (modalTop < 0) { modalTop = 0; }
        if (modalLeft < 0) { modalLeft = 0; }
        
        if( alwaysAtTop ) { 
        	windowElement.scrollTop(0); //Scroll to top of window
        } else {
        	modalTop += windowElement.scrollTop();
        	var spaceFromBottom = documentElement.height() - (modalTop + this.height());
        	if( spaceFromBottom < 0 && modalTop + spaceFromBottom >= 0)
        		modalTop += spaceFromBottom;
        }
        windowElement.scrollLeft(0);
        
        //Set the popup window to center
        this.css({'top': modalTop, 'left': modalLeft});
        
        return this;
	};
	
	$.fn.printModal = function() {
		window.print();
	};
	
	function resizeCurrentModal(e) {
		var alwaysAtTop = false;
		if( e.data.alwaysAtTop ) { alwaysAtTop = e.data.alwaysAtTop; }
		currentModal.resizeModal( alwaysAtTop );
	}
	
	function closeCurrentModal(e) {
		e.preventDefault(); //Prevents browser from following links
		currentModal.closeModal();
	}
	
	function printCurrentModal(e) {
		e.preventDefault(); //Prevents browser from following links
		currentModal.printModal();
	}
	function closeOnESC(e) {
		if( e.which == 27 ) {
			$(this).off('keyup', closeOnESC);
			closeCurrentModal(e);
		}
	}
})( jQuery );

/* Create modal box clones and click events */
jQuery(document).ready(function($){
	$('.chaos-modal-link').one('click', processModalLink);
	
	function processModalLink(e) {
		var thisElement = $(this);
		
		e.preventDefault(); //Prevents browser from following links
		thisElement.off('click', processModalLink); //Remove content processing event. It should only be run once per link.
		
		var modalContent = thisElement.parent().find('.chaos-modal-box').first(),
			modalContentClone = false,
			imageRegex = /\.(jpeg|jpg|gif|png|bmp|wbmp)$/i,
			imageUrl = false;
		
		//Check for pre-defined modal content
		if( modalContent.length ) {
			modalContentClone = $('<div></div>').css({'padding': '20px'});
			modalContent.clone().appendTo(modalContentClone);
			modalContentClone = $('<div></div>').css({'background': '#fff'}).append(modalContentClone);
		} else {
			if( imageRegex.test(thisElement.attr('href')) ) {
				//Use link href as image url
				imageUrl = thisElement.attr('href');
			} else {
				//Check for single image inside of modal link
				var modalImage = thisElement.find('img');
				if( modalImage.length == 1 ) {
					if( imageRegex.test(modalImage.attr('src')) )
						imageUrl = modalImage.attr('src');
				}
			}
			
			if( imageUrl ) {
				//Send HEAD request to validate image url
				testRequest = $.ajax(imageUrl, {type: 'HEAD', async: false});
				if( testRequest.status == 200 ) {
					modalContentClone = $('<div></div>').css({'padding': '20px'});
					$('<img src="'+imageUrl+'" />').css({'display': 'block'}).appendTo(modalContentClone);
					modalContentClone = $('<div></div>').css({'background': '#fff'}).append(modalContentClone);
				}
			}
		}
		
		//Setup openModal event if content is available
		if( modalContentClone ) {
			//Assign unique ids to each modal link and content
			thisElement.attr({
				href: '#',
				id: 'chaos-modal-link-'+$.chaosModalIndex
			});
			modalContentClone.attr({
				id: 'chaos-modal-box-'+$.chaosModalIndex
			});
			modalContentClone.hide();
			
			//Append hidden modal content to body
			modalContentClone.appendTo('body');
			
			thisElement.click(function() {
				var thisId = $(this).attr('id'),
					indexRegex = /chaos-modal-link-(\d+)/i,
					chaosModalIndex = thisId.match(indexRegex);
				if( chaosModalIndex[1] ) {
					var modalBox = $('#chaos-modal-box-'+chaosModalIndex[1]);
					if( modalBox.length == 1 ) {
						modalBox.openModal();
					}
				}
				return false;
			});
			thisElement.click();
			
			$.chaosModalIndex++;
		}
		return false;
	}
});

jquery.modal.js
/*
 * jQuery Chaos Modal
 * By Matthew Sigley
 * Based on concept work by Kevin Liew - http://www.queness.com/post/77/simple-jquery-modal-window-tutorial
 * Version 1.5.0
 */

(function( $ ) {

	//Bowser IE/Edge version detection
	function getIEEdgeVersion() {
		var ua = typeof navigator !== 'undefined' ? navigator.userAgent : '';
		if( /msie|trident/i.test(ua) ) {
			var match = ua.match(/(?:msie |rv:)(\d+(\.\d+)?)/i);
    	return (match && match.length > 1 && match[1]) || '';
    } else if( /chrome.+? edge/i.test(ua) ) {
			var match = ua.match(/edge\/(\d+(\.\d+)?)/i);
    	return (match && match.length > 1 && match[1]) || '';
		}
    return '';
  }

	var windowElement = $(window),
		 IEEdgeVersion = '';
	
	if( bowser.msie )
		IEEdgeVersion = bowser.version;
	else if( bowser.msedge )
		IEEdgeVersion = bowser.version;
	else
		IEEdgeVersion = getIEEdgeVersion();
	
	//Detects images that are not visible
	var isVisible = function( element ) {
		if( element.offsetParent === null )
			return false;
		if( IEEdgeVersion < 11 ) {
			var computedDisplay = '';
			if( typeof getComputedStyle != 'undefined' )
				computedDisplay = window.getComputedStyle(element).display;
			else if( typeof element.currentStyle != 'undefined' )
				computedDisplay = element.currentStyle['display'];
			if( 'none' == computedDisplay || 'fixed' == computedDisplay || 'absolute' == computedDisplay )
				return false;
		}
		return true;
	};


	//Global variables for generating modal box clones
	$.chaosModalLoading  = null;
	$.modalMask          = null;
	$.chaosModalCurrent  = null;
	$.chaosModalMaxWidth = 0;
	$.chaosModalIndex    = 0;
	$.chaosModalImagesLoaded  = false;
	$.chaosModalIframesLoaded = false;
	
	$.fn.openModal = function( options ) {
		//If modal already open, return
		if($.chaosModalCurrent != null){
			return;
		}
		//Default options
		var defaults = { maxWidth: 960, closeLink: true, printLink: false, alwaysAtTop: false, preprocessing: false };
		    options  = $.extend({}, defaults, options);
		//Lazy Load iFrame content
		function iframeDataSrc() {
			var thisElement = $(this),
				iframeDataSrc = thisElement.data('src');
			if( iframeDataSrc )
				thisElement.attr('src', iframeDataSrc).attr('data-src', '');
		}
		this.find('iframe[src=""]').each(iframeDataSrc);

		//Clone modal content
		var clone           = this.clone(true),
			  bodyElement     = $('body'),
			  htmlElement     = $('html'),
			  documentElement = $(document),
			  windowElement   = $(window),
			  closeLink       = options['closeLink'],
			  printLink       = options['printLink'],
			  alwaysAtTop     = options['alwaysAtTop'];
		//Update $.chaosModalMaxWidth
		$.chaosModalMaxWidth = options['maxWidth'];
		
		//Update $.chaosModalCurrent
		$.chaosModalCurrent = clone;
		clone.attr({id: 'chaos-current-modal'});
		
		//Write print link if none exist
		if(clone.find('.print-link').length == 0 && printLink) {
			clone.prepend('<a class="print-link">Print</a>');
			clone.children('.print-link').css({'float': 'right', 'margin': '5px'});
		}
		
		//Write close link if none exist
		if(clone.find('.close-link').length == 0 && closeLink) {
			clone.prepend('<a class="close-link">Close</a>');
			clone.children('.close-link').css({'float': 'right', 'margin': '5px'});
		}
               
    //Set the popup window css
    clone.css({'display': 'block', 'position': 'absolute', 'background': '#fff', 'z-index': '9002', 'left': '-10000px', 'margin': '0', 'padding': '0'});
    clone.prependTo(bodyElement);
     
    var showModal = function() {
			//Lock popup window width
	    if(clone.width() > $.chaosModalMaxWidth){
	      clone.width($.chaosModalMaxWidth);
	    } else {
	      clone.width(clone.width());
	    }
	       	
	    //Fix content widths
	    clone.find('img').css('max-width', '100%'); 
	       	
	    clone.hide();
	       	
			//Calculate the modal mask size and popup window position
			clone.resizeModal( alwaysAtTop );
	     	
	    //transition effect
			$.modalMask.stop();
	    clone.fadeIn(2000, function() {
	      //Bind the window resize event
				clone.resizeModal( alwaysAtTop );
	      windowElement.bind('resize', {'alwaysAtTop': alwaysAtTop}, resizeCurrentModal);
				$.fn.closeLoading();
	    });
	        
	    //Bind the print link events if any close links exist
	    if(clone.find('.print-link').length > 0) {
	      clone.find('.print-link').bind('click', printCurrentModal);
	    }
	        
	    //Bind the close link events if any close links exist
	    if(clone.find('.close-link').length > 0) {
	      clone.find('.close-link').bind('click', closeCurrentModal);
	      }
	        
    };
        
		var isImageNotLoaded = function( index, img ) {
			if (!isVisible(img))
				return false;
			if (img.complete)
				return false;
			if (typeof img.naturalWidth == "undefined" || img.naturalWidth != 0)
				return false;
			return true;
		};
		
		var isIframeNotLoaded = function( index, iframe ) {
			if( IEEdgeVersion ) //Ignore Iframes in IE due to security restrictions.
				return false;
			if (!isVisible(iframe))
		    	return false;
			var iframeDocument = iframe.contentDocument || iframe.contentWindow.document
			if ('complete' == iframeDocument.readyState)
		    return false;
			return true;
		};

    //Show modal after all images and iframes have loaded
    var images  = clone.find('img').filter(isImageNotLoaded),
        iframes = clone.find('iframe').filter(isIframeNotLoaded);
				$.imagesToLoad  = images.length;
				$.iframesToLoad = iframes.length;

    if(!$.imagesToLoad && !$.iframesToLoad)
      showModal();
    else {
      images.on('load', function() {
        $.imagesToLoad -= 1;
        if(!$.imagesToLoad && !$.iframesToLoad)
        	showModal();
      });
      iframes.on('load', function() {
        $.iframesToLoad -= 1;
        if(!$.imagesToLoad && !$.iframesToLoad)
        	showModal();
      });
    }
        
    clone.show();
        
    return this;
	};
	$.fn.closeModal = function() {
		//Unbind the window resize event
		$(window).unbind('resize', resizeCurrentModal);
		//Clear $.chaosModalCurrent
		$.chaosModalCurrent = null;
		
		//Unbind the close link events if any close links exist
        if(this.find('.close-link').length > 0) {
        	this.find('.close-link').unbind('click');
        }
		
		//Remove the popup window
		this.hide().remove();
		
		//Remove the mask div
		$('#chaos-modal-mask').hide().remove();
		$.fn.closeLoading();
		
		return this;
	};
	
	$.fn.resizeModal = function( alwaysAtTop ) {
		var documentElement = $(document),
			  windowElement   = $(window),
			  htmlElement     = $('html'),
			  maskHeight      = documentElement.height(), //Get the screen height and width
			  maskWidth       = htmlElement.width(),
			  winH            = windowElement.height(), //Get the window height and width
        winW            = windowElement.width();
        
        //Check for invalid mask dimensions
        if(maskHeight < this.height()) { maskHeight = this.height(); }
        if(maskWidth < winW) { maskWidth = winW; }
        
        //Set height and width to mask to fill up the whole screen
        $('#chaos-modal-mask').css({'width':maskWidth,'height':maskHeight});
                
        //Calculate popup window position
        var modalTop  = winH/2-this.height()/2,
        	   modalLeft = winW/2-this.width()/2;
        
        //Check for invalid window positions
        if (modalTop < 0) { modalTop = 0; }
        if (modalLeft < 0) { modalLeft = 0; }
        
        if( alwaysAtTop ) { 
        	windowElement.scrollTop(0); //Scroll to top of window
        } else {
        	modalTop += windowElement.scrollTop();
        	var spaceFromBottom = documentElement.height() - (modalTop + this.height());
        	if( spaceFromBottom < 0 && modalTop + spaceFromBottom >= 0)
        		modalTop += spaceFromBottom;
        }
        windowElement.scrollLeft(0);
        
        //Set the popup window to center
        this.css({'top': modalTop, 'left': modalLeft});
        
        return this;
	};
	
	$.fn.printModal = function() {
		window.print();
	};
	
	function resizeCurrentModal(e) {
		var alwaysAtTop = false;
		if( e.data.alwaysAtTop ) { alwaysAtTop = e.data.alwaysAtTop; }
		$.chaosModalCurrent.resizeModal( alwaysAtTop );
	}
	
	function closeCurrentModal(e) {
		e.preventDefault(); //Prevents browser from following links
		//Velocity control
		if(Date.now() - $.openTime < 500){
			return false;
		}
		$(document).off('.chaosmodal');
		if($.chaosModalCurrent != null)
			$.chaosModalCurrent.closeModal();
		else{
			$(window).trigger('chaos-modal-preprocess-interrupt');
			$('#chaos-modal-mask').hide().remove();
			$.fn.closeLoading();
		}
		$.modalMask = null;
	}
	
	function printCurrentModal(e) {
		e.preventDefault(); //Prevents browser from following links
		$.chaosModalCurrent.printModal();
	}
	function closeOnESC(e) {
		if( e.which == 27 ) {
			closeCurrentModal(e);
		}
	}

	//Function shows the loading screen (modalMask and loading gif)
	$.fn.showLoading = function(){
		//Sets open time for velocity control
		$.openTime = Date.now();
		if($.modalMask != null){
			return;
		}
		var bodyElement   = $('body'),
		    windowElement = $(window),
		    winH          = windowElement.height(), //Get the window height and width
		    winW          = windowElement.width();
			  maskClose     = true;
		//Create mask div
		$.modalMask = $('<div id="chaos-modal-mask" class="chaos-modal-mask"></div>');
		$.modalMask.css({'position': 'absolute', 'z-index': '9000', 'background-color': '#000', 'display': 'none', 'top': '0', 'left': '0'});
		
		$.modalMask.prependTo(bodyElement);

		//Mask transition effect    
	   $.modalMask.fadeIn(500);   
	   $.modalMask.fadeTo("fast",0.8);

		var documentElement = $(document),
			  windowElement   = $(window),
			  htmlElement     = $('html'),
			  maskHeight      = documentElement.height(), //Get the screen height and width
			  maskWidth       = htmlElement.width();
		
		//Bind close event to mask click and when the ESC key is pressed
		if(maskClose) {
			$.modalMask.on('click', closeCurrentModal);
			documentElement.on('keyup.chaosmodal', closeOnESC);
		}

		//Check for invalid mask dimensions
    if(maskHeight < this.height()) { maskHeight = this.height(); }
    if(maskWidth < winW) { maskWidth = winW; }
        
    //Set height and width to mask to fill up the whole screen
    $('#chaos-modal-mask').css({'width':maskWidth,'height':maskHeight});

		//Sets positioning for the loading bar
		
		$('body').append('<div id="modal-loading" style="position: fixed; z-index: 9001; color: #fff;">Loading</div>');
		
		//Center Loading
		posx = (winW / 2) - parseInt($('#modal-loading').css('width')) / 2;
		posy = (winH / 2) - parseInt($('#modal-loading').css('height')) / 2;
		
		$.chaosModalLoading = $('#modal-loading');
		$.chaosModalLoading.css('top', posy);
		$.chaosModalLoading.css('left', posx);
		var alwaysAtTop = true;
  	windowElement.bind('resize', {'alwaysAtTop': alwaysAtTop}, resizeLoading);	
	}
	function resizeLoading(){
		var documentElement = $(document),
			  windowElement   = $(window),
			  htmlElement     = $('html'),
			  maskHeight      = documentElement.height(), //Get the screen height and width
			  maskWidth       = htmlElement.width(),
			  winH            = windowElement.height(), //Get the window height and width
        winW            = windowElement.width();

		$('#chaos-modal-mask').css({'width':maskWidth,'height':maskHeight});

		//Center Loading
		posx = (winW / 2) - parseInt($('#modal-loading').css('width')) / 2;
		posy = (winH / 2) - parseInt($('#modal-loading').css('height')) / 2;
		
		$.chaosModalLoading = $('#modal-loading');
		$.chaosModalLoading.css('top', posy);
		$.chaosModalLoading.css('left', posx);
	}

	//Closes the loading bar
	$.fn.closeLoading = function(){
		$.chaosModalLoading.remove();
	}
	
})( jQuery );

/* Create modal box clones and click events */
jQuery(document).ready(function($){
	//Fix css classes
	$('.chaos-modal-link').not('a').each( function() {
		var thisElement = $(this),
			childElements = thisElement.children();
		thisElement.removeClass('chaos-modal-link');
		if( childElements.not('a').length )
			return;
		if( childElements.length == 1 )
			childElements.addClass('chaos-modal-link');
	});
	
	//Attach click event to links
	$('a.chaos-modal-link').on('click', processModalLink);
	
	//Lazy Load iFrame content
	function iframeDataSrc() {
		var thisElement = $(this),
			iframeSrc = thisElement.attr('src');
		thisElement.attr('data-src', iframeSrc).attr('src', '');
	}
	$('.chaos-modal-box:hidden iframe').each(iframeDataSrc);
	$('.chaos-modal-link').each(function() {
		var modalContentId = $(this).data('chaos-modal-box-id');
		if( modalContentId )
			$('#'+modalContentId+':hidden iframe').each(iframeDataSrc);
	});
	
	function processModalLink(e) {
		var thisElement = $(this);
	
		e.preventDefault(); //Prevents browser from following links
		thisElement.off('click', processModalLink); //Remove content processing event. It should only be run once per link.
		
		var modalContentId    = thisElement.data('chaos-modal-box-id'),
				modalContent      = false,
				modalContentClone = false,
				imageRegex        = /\.(jpeg|jpg|gif|png|bmp|wbmp)$/i,
				imageUrl          = false;
		
		if( modalContentId )
			modalContent = $('#'+modalContentId).first();
		else
			modalContent = thisElement.parent().find('.chaos-modal-box').first();
		
		//Check for pre-defined modal content
		if( modalContent.length ) {
			modalContentClone = $('<div></div>').css({'padding': '20px'});
			modalContent.clone(true).css({'display': 'block'}).appendTo(modalContentClone);
			modalContentClone = $('<div></div>').css({'background': '#fff'}).append(modalContentClone);
		} else {
			if( imageRegex.test(thisElement.attr('href')) ) {
				//Use link href as image url
				imageUrl = thisElement.attr('href');
			} else {
				//Check for single image inside of modal link
				var modalImage = thisElement.find('img');
				if( modalImage.length == 1 ) {
					if( imageRegex.test(modalImage.attr('src')) )
						imageUrl = modalImage.attr('src');
				}
			}
			
			if( imageUrl ) {
				//Send HEAD request to validate image url
				testRequest = $.ajax(imageUrl, {type: 'HEAD', async: false});
				if( testRequest.status == 200 ) {
					modalContentClone = $('<div></div>').css({'padding': '20px'});
					$('<img src="'+imageUrl+'" />').css({'display': 'block'}).appendTo(modalContentClone);
					modalContentClone = $('<div></div>').css({'background': '#fff'}).append(modalContentClone);
				}
			}
		}
		//Setup openModal event if content is available
		if( modalContentClone ) {
			//Remove original content element to avoid id conflicts
			modalContent.remove();
			//Assign unique ids to each modal link and content
			thisElement.attr({
				href: '#',
				id: 'chaos-modal-link-'+$.chaosModalIndex
			});
			modalContentClone.attr({
				id: 'chaos-modal-box-'+$.chaosModalIndex
			});
			modalContentClone.hide();
			
			//Append hidden modal content to body
			modalContentClone.appendTo('body');
			thisElement.click(function() {
				$.fn.showLoading();
				//Get options from modal-link
				
				var options = { maxWidth : 960,
									 closeLink: true,
									 printLink: false, 
									 alwaysAtTop: false, 
									 preprocessing: false };
				
				$.each(options, function(key, value){
					if(typeof(thisElement.data('chaosmodal-'+key))!== 'undefined'){
						console.log(typeof(thisElement.data('chaosmodal-'+key)));
						if(key == 'maxWidth'){
							options[key] = parseInt(thisElement.data('chaosmodal-'+key));
						} else {
							options[key] = (thisElement.data('chaosmodal-'+key) == 'true');
						}
					}
				});

				preprocess = thisElement.attr('data-preprocess');
				var thisId = $(this).attr('id'),
					indexRegex = /chaos-modal-link-(\d+)/i,
					chaosModalIndex = thisId.match(indexRegex);
				if( chaosModalIndex[1] ) {
					var modalBox = $('#chaos-modal-box-'+chaosModalIndex[1]);
					if( modalBox.length == 1 && preprocess != 'true' ) {
						modalBox.openModal(options);
					} else if (modalBox.length == 1 && preprocess == 'true'){
						thisElement.trigger("chaos-modal-preprocess", modalBox);
					}
				}
				return false;
			});
			thisElement.click();
			
			$.chaosModalIndex++;
		}
		return false;
	}
	//Polyfill for Date.now()
	if(!Date.now){
			Date.now = function now(){
				return new Date().getTime();
			};
	}
});